modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -817,5 +817,1068 @@ int HashTSDFVolumeCPU::getVisibleBlocks(int currFrameId, int frameThreshold) con
+                                {
+                                    Point3f point = base_point + volume.voxelCoordToVolume(voxelIdx);
+                                    localPoints.push_back(toPtype(point));
+                                    if (needNormals)
+                                    {
+                                        Point3f normal = volume._getNormalVoxel(point);
+                                        //Point3f normal(0,0,0);
+                                        localNormals.push_back(toPtype(normal));
+                                    }
+                                }
+                            }
+
+                    AutoLock al(mutex);
+                    //_pVecs.push_back(localPoints);
+                    //_nVecs.push_back(localNormals);
+                    pVecs.push_back(localPoints);
+                    nVecs.push_back(localNormals);
+                }
+            }
+        };
+
+        parallel_for_(_fetchRange, _HashFetchPointsNormalsInvoker, nstripes);
+
+
+
+        std::vector<ptype> points, normals;
+        for (size_t i = 0; i < pVecs.size(); i++)
+        {
+            points.insert(points.end(), pVecs[i].begin(), pVecs[i].end());
+            normals.insert(normals.end(), nVecs[i].begin(), nVecs[i].end());
+        }
+
+        _points.create((int)points.size(), 1, POINT_TYPE);
+        if (!points.empty())
+            Mat((int)points.size(), 1, POINT_TYPE, &points[0]).copyTo(_points.getMat());
+
+        if (_normals.needed())
+        {
+            _normals.create((int)normals.size(), 1, POINT_TYPE);
+            if (!normals.empty())
+                Mat((int)normals.size(), 1, POINT_TYPE, &normals[0]).copyTo(_normals.getMat());
+        }
+    }
+}
+
+void HashTSDFVolumeGPU::fetchNormals(InputArray _points, OutputArray _normals) const
+{
+    CV_TRACE_FUNCTION();
+
+    if (_normals.needed())
+    {
+        Points points = _points.getMat();
+        CV_Assert(points.type() == POINT_TYPE);
+        _normals.createSameSize(_points, _points.type());
+        Normals normals = _normals.getMat();
+        const HashTSDFVolumeGPU& _volume = *this;
+        auto HashPushNormals             = [&](const ptype& point, const int* position) {
+            const HashTSDFVolumeGPU& volume(_volume);
+            Affine3f invPose(volume.pose.inv());
+            Point3f p = fromPtype(point);
+            Point3f n = nan3;
+            if (!isNaN(p))
+            {
+                Point3f voxelPoint = invPose * p;
+                n                  = volume.pose.rotation() * volume._getNormalVoxel(voxelPoint);
+            }
+            normals(position[0], position[1]) = toPtype(n);
+        };
+        points.forEach(HashPushNormals);
+    }
+
+}
+
+int HashTSDFVolumeGPU::getVisibleBlocks(int currFrameId, int frameThreshold) const
+{
+    int numVisibleBlocks = 0;
+    //! TODO: Iterate over map parallely?
+    for (const auto& keyvalue : volumeUnits)
```

Comment:
1. +
2. Is this condition checked now for CPU code?

Reply:
Actually on second thought it is not required, since the `lastVisibleIdx` will be updated in the integrate step only if it is in the frustum. 
Perhaps this is why I didn't implement the check again here the first time around.